### PR TITLE
Set dependency with ensembl-py

### PR DESCRIPTION
**Pull request by JAlvarezJarreta**
## Description

The unit test framework was moved to `ensembl-py` some time ago, and some additional updates/bugfixes were done on that side, but we had not moved our code base and dependencies to start using that repository. It is finally here.

**Related JIRA tickets:**
- ENSCOMPARASW-4757

## Overview of changes
Most changes have been at the configuration level, since I have taken the opportunity to update all these as expected. The other minor changes have been to remove the bits that were transferred to `ensembl-py`, plus update docstrings/license headers to the agreed format.

## Testing
I have run several unit tests, even moved some database tests from `ensembl-py` to our repository, and everything worked as expected in Codon.

## Notes
- Any issues in CITest can be ignored since this project is on hold at the moment.
- Changes in our environment set up and installation will be required after this is merged.

### CODE REVIEW

File: src/python/tests/test_citest.py
Comment (CristiGuijarro) [inline src/python/tests/test_citest.py:38]: :cry:

**Review by CristiGuijarro — COMMENTED** [review]
I think I have reviewed this as best I can and compared with [ensembl-py PR#3](https://github.com/Ensembl/ensembl-py/pull/3) to check for any unintentional function loss :shrug: . I think everything I have put is simply a suggestion and not a "requirement" for approval.
Also can we rename `pyproject.toml`? The mention `project` in there makes it seem temporary and separate from the rest of the codebase - or is this the intention? I have no inspired preferences for a rename however.

File: .gitignore
Comment (JAlvarezJarreta) [reply .gitignore:1]: I have just followed the same idea as for `ensembl-py`. Excluding is always "easier" in my opinion, so you see quickly when you are picking something you shouldn't. But will wait for the second reviewer to make a decision.

File: .codecov.yml
Comment (JAlvarezJarreta) [reply .codecov.yml:1]: I have taken the opportunity to do some code clean-up in most config files (you know me :innocent:). Happy to revert it if you want :smile:

File: pyproject.toml
Comment (JAlvarezJarreta) [reply pyproject.toml:39]: I would need a really compelling case to disable these two since `with` is strongly suggested by Python as it handles the handler closure internally, and the `f-string` provides more readable formatting of strings and it has been proven to have better performance.

File: .gitignore
Comment (CristiGuijarro) [reply .gitignore:1]: It's a totally subjective thing, in both cases you could see quickly when you are not picking up something you should or picking up something you shouldn't! I suggest a git include rather than ignore, because that is less likely to change so much or be added to, and in this particular case, the anti-ignore list would be smaller than the to be ignored list. I am happy either way, but wanted to put forward the suggestion for consideration :nerd_face:

File: pyproject.toml
Comment (CristiGuijarro) [reply pyproject.toml:39]: There are 2 cases listed in the `feature/homology_annotation` branch for both ([here](https://github.com/Ensembl/ensembl-compara/blob/feature/homology_annotation/src/python/lib/ensembl/compara/runnable/SplitFasta.py#L97) and [here](https://github.com/Ensembl/ensembl-compara/blob/feature/homology_annotation/src/python/lib/ensembl/compara/runnable/SplitFasta.py#L102)). But on another note, I think with some vigilance, these can be picked up by the reviewer's eye if necessary, and in the team we can push for the usage.

**Comment by JAlvarezJarreta** [comment]
> I think I have reviewed this as best I can and compared with [ensembl-py PR#3](https://github.com/Ensembl/ensembl-py/pull/3) to check for any unintentional function loss shrug . I think everything I have put is simply a suggestion and not a "requirement" for approval. Also can we rename `pyproject.toml`? The mention `project` in there makes it seem temporary and separate from the rest of the codebase - or is this the intention? I have no inspired preferences for a rename however.

I'm afraid the name of `pyproject.toml` is fixed and cannot be changed, as it is the exact name searched by `pytest` and `mypy` (among others).

**Comment by JAlvarezJarreta** [comment]
`pylint` will fail for the HAL alignment module. I haven't addressed this error because it has been already taken care of on its own branch, so once it is merged into `master` this will be resolved. However, if you think it may be better to have Travis passing on `master`, I'm happy to silence it temporarily.

File: .codecov.yml
Comment (CristiGuijarro) [reply .codecov.yml:1]: I think this should be reverted, just because we are generally strict on this sort of change, no exceptions!

File: .gitignore
Comment (CristiGuijarro) [reply .gitignore:1]:
```suggestion
# Ignore everything
*
# Except these dirs
!conf/
!docs/
!modules/
!pipelines/
!scripts/
!sql/
!src/
!travisci/
!xs/
# And these files
!.codecov.yml
!cpanfile
!DEPRECATED.md
!.gitignore
!LICENSE
!.mypy.ini
!NOTICE
!perlcriticrc
!PIP_VERSION
!pull_request_template.md
!pylintrc
!README.md
!run_all_tests.sh
!setup.py
!.travis.yml
!yamllintrc
```

File: pyproject.toml
Comment (CristiGuijarro) [reply pyproject.toml:31]: I still think there are situations in which `consider-using-with` should not be enforced.

**Review by CristiGuijarro — COMMENTED** [review]
I think get rid of the lines to disable pylint for the hal liftover stuff before merging once proven the PR will pass Travis- perhaps you were going to do this anyway? It's a broad PR, so waiting on other reviewer to see if they  pick up stuff I have missed before approving, but otherwise :+1:

File: pyproject.toml
Comment (JAlvarezJarreta) [reply pyproject.toml:31]: So far Travis seems happy with this. When these come I will be happy to discuss it again :smile:

File: src/python/tests/conftest.py
Comment (twalsh-ebi) [inline src/python/tests/conftest.py:79]: Again out of curiosity, why does this use `shutil.copy` instead of the default `shutil.copy2`? Is it related to the need to have different timestamps between reference and target files?

**Review by twalsh-ebi — COMMENTED** [review]
There is a lot changing here, and inasmuch as I can follow what's going on, it looks good to me. I have just a handful of comments and questions.

As a gentle reminder, please be mindful of the [issue with .travis.yml, lines 28-34](https://github.com/Ensembl/ensembl-compara/blob/05b19a1bc8521dfe2df6efd5759f09c65f7a32d2/.travis.yml#L28-L34), which may necessitate some adjustments to the PR, or careful merging into the master branch.

**Review by CristiGuijarro — APPROVED** [review]

**Review by twalsh-ebi — APPROVED** [review]

File: .travis.yml
Comment (JAlvarezJarreta) [reply .travis.yml:175]: Yeah, sorry, you are completely right, it is not necessary unless we are planning on making changes on the installed repo (which we are not) :man_facepalming: Will revert this change.

File: src/python/tests/conftest.py
Comment (JAlvarezJarreta) [reply src/python/tests/conftest.py:79]: Yes, that is required to ensure `filecmp` methods are forced to inspect the content of the file even when the size matches.

**Pull request merged**
